File: apps/proof-lab/ios/DoughKit/Sources/DoughKit/Batch.swift
import Foundation

public struct StageProgress: Codable, Sendable, Equatable {
	/// Hours added or removed at the bench — "it's not ready, give it another 30".
	public var adjustmentHours: Double
	public var completedAt: Date?
	/// Set when you've seen the alert, which is what clears the badge.
	public var acknowledgedAt: Date?

	public init(adjustmentHours: Double = 0, completedAt: Date? = nil, acknowledgedAt: Date? = nil) {
		self.adjustmentHours = adjustmentHours
		self.completedAt = completedAt
		self.acknowledgedAt = acknowledgedAt
	}
}

/// Tasting notes for a prototype run. The point of the log is being able to tell why
/// batch 14 was better than batch 13.
public struct BatchReview: Codable, Sendable, Equatable {
	public var handling: Int
	public var extensibility: Int
	public var ovenSpring: Int
	public var crumb: Int
	public var flavor: Int
	public var crust: Int
	public var overall: Int
	public var wouldRepeat: Bool
	public var notes: String

	public init(
		handling: Int = 3,
		extensibility: Int = 3,
		ovenSpring: Int = 3,
		crumb: Int = 3,
		flavor: Int = 3,
		crust: Int = 3,
		overall: Int = 3,
		wouldRepeat: Bool = true,
		notes: String = ""
	) {
		self.handling = handling
		self.extensibility = extensibility
		self.ovenSpring = ovenSpring
		self.crumb = crumb
		self.flavor = flavor
		self.crust = crust
		self.overall = overall
		self.wouldRepeat = wouldRepeat
		self.notes = notes
	}

	public struct Axis: Sendable, Equatable, Identifiable {
		public let id: String
		public let name: String
		public let score: Int
	}

	public var axes: [Axis] {
		[
			Axis(id: "handling", name: "Handling", score: handling),
			Axis(id: "extensibility", name: "Extensibility", score: extensibility),
			Axis(id: "ovenSpring", name: "Oven spring", score: ovenSpring),
			Axis(id: "crumb", name: "Crumb", score: crumb),
			Axis(id: "flavor", name: "Flavour", score: flavor),
			Axis(id: "crust", name: "Crust", score: crust),
			Axis(id: "overall", name: "Overall", score: overall)
		]
	}
}

public enum StageStatus: String, Sendable {
	case upcoming
	case active
	case due
	case overdue
	case done
}

public struct Batch: Codable, Sendable, Equatable, Identifiable {
	public var id: UUID
	public var name: String
	public var createdAt: Date
	/// When the mix actually started. Everything downstream is derived from this.
	public var startAt: Date
	public var formula: DoughFormula
	public var plan: FermentationPlan
	public var mixerCapacityKg: Double
	public var progress: [String: StageProgress]
	public var review: BatchReview?
	public var notes: String
	public var tags: [String]
	public var isArchived: Bool

	public init(
		id: UUID = UUID(),
		name: String,
		createdAt: Date = Date(),
		startAt: Date,
		formula: DoughFormula,
		plan: FermentationPlan,
		mixerCapacityKg: Double = 20,
		progress: [String: StageProgress] = [:],
		review: BatchReview? = nil,
		notes: String = "",
		tags: [String] = [],
		isArchived: Bool = false
	) {
		self.id = id
		self.name = name
		self.createdAt = createdAt
		self.startAt = startAt
		self.formula = formula
		self.plan = plan
		self.mixerCapacityKg = mixerCapacityKg
		self.progress = progress
		self.review = review
		self.notes = notes
		self.tags = tags
		self.isArchived = isArchived
	}
}

public extension Batch {
	/// Absolute stage times, honouring adjustments and manual completions. Completing a
	/// stage early or late drags everything after it along.
	var timeline: [ScheduledStage] {
		var stages: [ScheduledStage] = []
		stages.reserveCapacity(plan.stages.count)
		var cursor = startAt

		for stage in plan.stages {
			let state = progress[stage.id]
			let planned = max(0, stage.hours + (state?.adjustmentHours ?? 0)) * 3600
			let end = state?.completedAt ?? cursor.addingTimeInterval(planned)
			let windowEnd = stage.usableWindowHours.map { end.addingTimeInterval($0 * 3600) }
			stages.append(
				ScheduledStage(
					id: stage.id,
					stage: stage,
					start: cursor,
					end: end,
					windowEnd: windowEnd,
					foldTimes: Scheduler.foldTimes(for: stage, start: cursor, end: end)
				)
			)
			cursor = end
		}
		return stages
	}

	var readyAt: Date {
		let stages = timeline
		guard let bakeIndex = stages.firstIndex(where: { $0.stage.kind == .bake }) else {
			return stages.last?.end ?? startAt
		}
		return bakeIndex > 0 ? stages[bakeIndex - 1].end : startAt
	}

	var fermentationLoadHours: Double { plan.fermentationLoadHours }

	/// Yeast the current schedule asks for, on an instant-dry basis.
	var suggestedInstantYeastPercent: Double {
		Leavening.instantYeastPercent(
			equivalentHours: fermentationLoadHours,
			saltPercent: formula.saltPercent,
			sugarPercent: formula.sugarPercent,
			prefermentedFlourFraction: formula.prefermentedFlourFraction
		)
	}

	var suggestedLevainPercent: Double {
		Leavening.levainPercent(
			equivalentHours: fermentationLoadHours,
			saltPercent: formula.saltPercent
		)
	}

	func status(of stage: ScheduledStage, at now: Date = Date()) -> StageStatus {
		if progress[stage.id]?.completedAt != nil { return .done }
		if now < stage.start { return .upcoming }
		if now < stage.end { return .active }
		if now >= stage.end.addingTimeInterval(30 * 60) { return .overdue }
		return .due
	}

	func currentStage(at now: Date = Date()) -> ScheduledStage? {
		let stages = timeline
		if let active = stages.first(where: { status(of: $0, at: now) == .active }) { return active }
		return stages.first { status(of: $0, at: now) != .done }
	}

	/// Stages whose timer has elapsed without being acknowledged. This is the badge count.
	func dueStages(at now: Date = Date()) -> [ScheduledStage] {
		timeline.filter { stage in
			guard stage.stage.alerts else { return false }
			let state = progress[stage.id]
			guard state?.completedAt == nil, state?.acknowledgedAt == nil else { return false }
			return now >= stage.end
		}
	}

	var isFinished: Bool {
		guard let last = plan.stages.last else { return true }
		return progress[last.id]?.completedAt != nil
	}
}

// MARK: - Alerts

public struct DoughAlert: Sendable, Equatable, Identifiable {
	public enum Kind: String, Sendable {
		case stageEnd
		case fold
		case windowOpen
		case windowClosing
		case prefermentReady
	}

	public let id: String
	public let batchID: UUID
	public let batchName: String
	public let stageID: String
	public let kind: Kind
	public let fireAt: Date
	public let title: String
	public let body: String
}

public extension Batch {
	/// Every alert this batch still owes you, in fire order.
	func upcomingAlerts(from now: Date = Date()) -> [DoughAlert] {
		var alerts: [DoughAlert] = []

		for stage in timeline {
			let state = progress[stage.id]
			guard state?.completedAt == nil else { continue }

			for (index, fold) in stage.foldTimes.enumerated() where fold > now {
				alerts.append(
					DoughAlert(
						id: "\(id.uuidString)|\(stage.id)|fold|\(index)",
						batchID: id,
						batchName: name,
						stageID: stage.id,
						kind: .fold,
						fireAt: fold,
						title: "Fold \(index + 1) — \(name)",
						body: "\(stage.stage.title): time for a set of folds."
					)
				)
			}

			guard stage.stage.alerts, stage.end > now else { continue }

			let kind: DoughAlert.Kind = stage.stage.kind == .preferment
				? .prefermentReady
				: (stage.stage.kind.isCold ? .windowOpen : .stageEnd)

			alerts.append(
				DoughAlert(
					id: "\(id.uuidString)|\(stage.id)|end",
					batchID: id,
					batchName: name,
					stageID: stage.id,
					kind: kind,
					fireAt: stage.end,
					title: "\(stage.stage.title) done — \(name)",
					body: Self.body(for: stage, kind: kind)
				)
			)

			if let windowEnd = stage.windowEnd, windowEnd > now {
				alerts.append(
					DoughAlert(
						id: "\(id.uuidString)|\(stage.id)|window",
						batchID: id,
						batchName: name,
						stageID: stage.id,
						kind: .windowClosing,
						fireAt: windowEnd,
						title: "Window closing — \(name)",
						body: "\(name) is at the end of its usable cold window. Use it or bin it."
					)
				)
			}
		}

		return alerts.sorted { $0.fireAt < $1.fireAt }
	}

	private static func body(for stage: ScheduledStage, kind: DoughAlert.Kind) -> String {
		switch kind {
		case .prefermentReady:
			"Preferment is ripe — domed and just starting to fall. Mix now."
		case .windowOpen:
			"Cold ferment is ready. Usable for another \(Formatting.hours(stage.stage.usableWindowHours ?? 0))."
		default:
			stage.stage.detail.isEmpty ? "Timer's up on \(stage.stage.title.lowercased())." : stage.stage.detail
		}
	}
}

/// Assigns cumulative badge numbers across every batch, so the icon badge is correct even
/// if the app never runs between alerts.
public enum AlertScheduler {
	/// iOS keeps at most 64 pending local notifications per app.
	public static let pendingLimit = 60

	public static func badgedAlerts(
		batches: [Batch],
		from now: Date = Date(),
		startingBadge: Int = 0,
		limit: Int = pendingLimit
	) -> [(alert: DoughAlert, badge: Int)] {
		let all = batches
			.filter { !$0.isArchived }
			.flatMap { $0.upcomingAlerts(from: now) }
			.sorted { $0.fireAt < $1.fireAt }
			.prefix(limit)

		var badge = max(0, startingBadge)
		return all.map { alert in
			// Fold reminders are nudges, not outstanding work; they don't raise the badge.
			if alert.kind != .fold { badge += 1 }
			return (alert, badge)
		}
	}

	/// Badge to show right now, counting only alerts that have already fired unacknowledged.
	public static func currentBadge(batches: [Batch], at now: Date = Date()) -> Int {
		batches.filter { !$0.isArchived }.reduce(0) { $0 + $1.dueStages(at: now).count }
	}
}

File: apps/proof-lab/doughcore/src/main/kotlin/com/blahmonster/prooflab/core/Batch.kt
package com.blahmonster.prooflab.core

import kotlin.math.max
import kotlinx.serialization.Serializable

@Serializable
data class StageProgress(
	/** Hours added or removed at the bench — "it's not ready, give it another 30". */
	val adjustmentHours: Double = 0.0,
	val completedAt: Long? = null,
	/** Set when you've seen the alert, which is what clears the badge. */
	val acknowledgedAt: Long? = null,
)

/**
 * Tasting notes for a prototype run. The point of the log is being able to tell why batch 14
 * was better than batch 13.
 */
@Serializable
data class BatchReview(
	val handling: Int = 3,
	val extensibility: Int = 3,
	val ovenSpring: Int = 3,
	val crumb: Int = 3,
	val flavor: Int = 3,
	val crust: Int = 3,
	val overall: Int = 3,
	val wouldRepeat: Boolean = true,
	val notes: String = "",
) {
	data class Axis(val id: String, val name: String, val score: Int)

	val axes: List<Axis>
		get() = listOf(
			Axis("handling", "Handling", handling),
			Axis("extensibility", "Extensibility", extensibility),
			Axis("ovenSpring", "Oven spring", ovenSpring),
			Axis("crumb", "Crumb", crumb),
			Axis("flavor", "Flavour", flavor),
			Axis("crust", "Crust", crust),
			Axis("overall", "Overall", overall),
		)
}

enum class StageStatus { UPCOMING, ACTIVE, DUE, OVERDUE, DONE }

@Serializable
data class Batch(
	val id: String,
	val name: String,
	val createdAt: Long,
	/** When the mix actually started. Everything downstream is derived from this. */
	val startAt: Long,
	val formula: DoughFormula,
	val plan: FermentationPlan,
	val mixerCapacityKg: Double = 20.0,
	val progress: Map<String, StageProgress> = emptyMap(),
	val review: BatchReview? = null,
	val notes: String = "",
	val tags: List<String> = emptyList(),
	val isArchived: Boolean = false,
) {
	/**
	 * Absolute stage times, honouring adjustments and manual completions. Completing a stage
	 * early or late drags everything after it along.
	 */
	val timeline: List<ScheduledStage>
		get() {
			val stages = mutableListOf<ScheduledStage>()
			var cursor = startAt
			for (stage in plan.stages) {
				val state = progress[stage.id]
				val planned = hoursToMillis(stage.hours + (state?.adjustmentHours ?: 0.0))
				val end = state?.completedAt ?: (cursor + planned)
				val windowEnd = stage.usableWindowHours?.let { end + hoursToMillis(it) }
				stages.add(
					ScheduledStage(
						id = stage.id,
						stage = stage,
						start = cursor,
						end = end,
						windowEnd = windowEnd,
						foldTimes = Scheduler.foldTimes(stage, cursor, end),
					),
				)
				cursor = end
			}
			return stages
		}

	val readyAt: Long
		get() {
			val stages = timeline
			val bakeIndex = stages.indexOfFirst { it.stage.kind == StageKind.BAKE }
			if (bakeIndex < 0) return stages.lastOrNull()?.end ?: startAt
			return if (bakeIndex > 0) stages[bakeIndex - 1].end else startAt
		}

	val fermentationLoadHours: Double get() = plan.fermentationLoadHours

	/** Yeast the current schedule asks for, on an instant-dry basis. */
	val suggestedInstantYeastPercent: Double
		get() = Leavening.instantYeastPercent(
			equivalentHours = fermentationLoadHours,
			saltPercent = formula.saltPercent,
			sugarPercent = formula.sugarPercent,
			prefermentedFlourFraction = formula.prefermentedFlourFraction,
		)

	val suggestedLevainPercent: Double
		get() = Leavening.levainPercent(fermentationLoadHours, formula.saltPercent)

	fun status(stage: ScheduledStage, now: Long): StageStatus = when {
		progress[stage.id]?.completedAt != null -> StageStatus.DONE
		now < stage.start -> StageStatus.UPCOMING
		now < stage.end -> StageStatus.ACTIVE
		now >= stage.end + 30 * 60_000L -> StageStatus.OVERDUE
		else -> StageStatus.DUE
	}

	fun currentStage(now: Long): ScheduledStage? {
		val stages = timeline
		return stages.firstOrNull { status(it, now) == StageStatus.ACTIVE }
			?: stages.firstOrNull { status(it, now) != StageStatus.DONE }
	}

	/** Stages whose timer has elapsed without being acknowledged. This is the badge count. */
	fun dueStages(now: Long): List<ScheduledStage> = timeline.filter { stage ->
		if (!stage.stage.alerts) return@filter false
		val state = progress[stage.id]
		state?.completedAt == null && state?.acknowledgedAt == null && now >= stage.end
	}

	val isFinished: Boolean
		get() {
			val last = plan.stages.lastOrNull() ?: return true
			return progress[last.id]?.completedAt != null
		}

	/** Every alert this batch still owes you, in fire order. */
	fun upcomingAlerts(now: Long): List<DoughAlert> {
		val alerts = mutableListOf<DoughAlert>()

		for (stage in timeline) {
			val state = progress[stage.id]
			if (state?.completedAt != null) continue

			stage.foldTimes.forEachIndexed { index, fold ->
				if (fold > now) {
					alerts.add(
						DoughAlert(
							id = "$id|${stage.id}|fold|$index",
							batchId = id,
							batchName = name,
							stageId = stage.id,
							kind = DoughAlert.Kind.FOLD,
							fireAt = fold,
							title = "Fold ${index + 1} — $name",
							body = "${stage.stage.title}: time for a set of folds.",
						),
					)
				}
			}

			if (!stage.stage.alerts || stage.end <= now) continue

			val kind = when {
				stage.stage.kind == StageKind.PREFERMENT -> DoughAlert.Kind.PREFERMENT_READY
				stage.stage.kind.isCold -> DoughAlert.Kind.WINDOW_OPEN
				else -> DoughAlert.Kind.STAGE_END
			}

			alerts.add(
				DoughAlert(
					id = "$id|${stage.id}|end",
					batchId = id,
					batchName = name,
					stageId = stage.id,
					kind = kind,
					fireAt = stage.end,
					title = "${stage.stage.title} done — $name",
					body = bodyFor(stage, kind),
				),
			)

			val windowEnd = stage.windowEnd
			if (windowEnd != null && windowEnd > now) {
				alerts.add(
					DoughAlert(
						id = "$id|${stage.id}|window",
						batchId = id,
						batchName = name,
						stageId = stage.id,
						kind = DoughAlert.Kind.WINDOW_CLOSING,
						fireAt = windowEnd,
						title = "Window closing — $name",
						body = "$name is at the end of its usable cold window. Use it or bin it.",
					),
				)
			}
		}

		return alerts.sortedBy { it.fireAt }
	}

	private fun bodyFor(stage: ScheduledStage, kind: DoughAlert.Kind): String = when (kind) {
		DoughAlert.Kind.PREFERMENT_READY ->
			"Preferment is ripe — domed and just starting to fall. Mix now."
		DoughAlert.Kind.WINDOW_OPEN ->
			"Cold ferment is ready. Usable for another " +
				"${Formatting.hours(stage.stage.usableWindowHours ?: 0.0)}."
		else ->
			if (stage.stage.detail.isEmpty()) "Timer's up on ${stage.stage.title.lowercase()}."
			else stage.stage.detail
	}
}

data class DoughAlert(
	val id: String,
	val batchId: String,
	val batchName: String,
	val stageId: String,
	val kind: Kind,
	val fireAt: Long,
	val title: String,
	val body: String,
) {
	enum class Kind { STAGE_END, FOLD, WINDOW_OPEN, WINDOW_CLOSING, PREFERMENT_READY }
}

/**
 * Assigns cumulative badge numbers across every batch, so the launcher badge is correct even
 * if the app never runs between alerts.
 */
object AlertScheduler {
	const val PENDING_LIMIT = 60

	fun badgedAlerts(
		batches: List<Batch>,
		now: Long,
		startingBadge: Int = 0,
		limit: Int = PENDING_LIMIT,
	): List<Pair<DoughAlert, Int>> {
		val all = batches
			.filterNot { it.isArchived }
			.flatMap { it.upcomingAlerts(now) }
			.sortedBy { it.fireAt }
			.take(limit)

		var badge = max(0, startingBadge)
		return all.map { alert ->
			// Fold reminders are nudges, not outstanding work; they don't raise the badge.
			if (alert.kind != DoughAlert.Kind.FOLD) badge += 1
			alert to badge
		}
	}

	/** Badge to show right now, counting only alerts that have already fired unacknowledged. */
	fun currentBadge(batches: List<Batch>, now: Long): Int =
		batches.filterNot { it.isArchived }.sumOf { it.dueStages(now).size }
}
